src/js/taskManager.js

```diff
@@ -0,0 +1,64 @@
+// eslint-disable-next-line import/no-cycle
+import tasks from './tasks.js';
+
+class TaskManager {
+  constructor() {
+    this.enterTodo = document.getElementById('enter-todo');
+    this.data = null;
+  }
+
+  addTask(value) {
+    const task = {
+      description: value,
+      completed: false,
+      // index: tasks.todoData.length + 1,
+    };
+    tasks.todoData = this.writeLocalCollection(tasks.storageKey, task);
+  }
+
+  deleteTask(taskIndex) {
+    const taskToDelete = tasks.todoData[taskIndex];
+    tasks.todoData = tasks.todoData.filter((data, index) => index !== taskIndex);
+    this.writeLocalCollection(tasks.storageKey, tasks.todoData, false);
+    return taskToDelete;
+  }
+
+  modifyTask(taskObject, taskIndex) {
+    // To receive only the keys necessary to be changed
+    Object.entries(taskObject).forEach(([key, value]) => {
+      tasks.todoData[taskIndex][key] = value;
+    });
+    this.writeLocalCollection(tasks.storageKey, tasks.todoData, false);
+  }
+
+  clearCompletedTasks(taskIndexes) {
+    tasks.todoData = tasks.todoData.filter((task, index) => !taskIndexes.includes(index));
+    this.writeLocalCollection(tasks.storageKey, tasks.todoData, false);
+  }
+
+  setValue(element, value = '') {
+    element.value = value;
+    return this.undefined; // Return undefined as though there's no return statement
+  }
+
+  readLocalCollection(key) {
+    this.data = JSON.parse(localStorage.getItem(key)) || [];
+    return this.data;
+  }
+
+  writeLocalCollection(key, item, single = true) { // When single is false, do a replace
+    let existingData;
+    if (single) {
+      existingData = this.readLocalCollection(key);
+      existingData.push(item);
+    } else {
+      existingData = item;
+    }
+
+    localStorage.setItem(key, JSON.stringify(existingData));
+    return existingData;
+  }
+}
+
+const taskManager = new TaskManager();
+export default taskManager;
```

Comment:
Here are some suggestions to improve the code and make it follow the DRY (Don't Repeat Yourself) and KISS (Keep It Simple, Stupid) principles:

Use a template literal to create the markup for each todo item, instead of manually concatenating HTML strings. This makes the code more readable and easier to maintain.
python
Copy code
```
const markupTodoItem = (data, index) => `
  <li class="todo-item">
    <div class="todo-entry" id="entry_${index}">
      <input type="checkbox" ${data.completed ? 'checked' : ''} class="status" title="Check!">
      <input type="text" class="todo" value="${data.description}">
      <i class="fa-solid fa-trash-can delete hide" title="Delete entry"></i>
      <i class="fa-solid fa-ellipsis-vertical reorder" title="Move"></i>
    </div>
    <hr class="divide">
  </li>
`;
```
Use the map() method instead of forEach() to create an array of todo items markup, and then join the array to create the final markup. This eliminates the need to declare an empty string variable and manually concatenate each todo item markup.
javascript
Copy code
```
const markupAllLists = () => {
  const allLists = tasks.todoData.map((data, index) => markupTodoItem(data, index));
  return allLists.join('');
};
```
Instead of re-rendering the entire list every time a task is deleted or modified, only update the affected todo item. This improves performance and reduces unnecessary DOM updates.
javascript
Copy code
const renderTask = (index) => {
  const todoEntry = document.getElementById(`entry_${index}`);
  const data = tasks.todoData[index];
  todoEntry.innerHTML = markupTodoItem(data, index);

  // add event listeners to the new todo entry
  const descInput = todoEntry.querySelector('.todo');
  const deleteButton = todoEntry.querySelector('.delete');
  const status = todoEntry.querySelector('.status');

  // ...
};

```
const addRefreshingListeners = () => {
  // ...

  descInput.forEach((element) => {
    ['focus', 'blur', 'keyup'].forEach((evt) => {
      element.addEventListener(evt, (e) => {
        const taskIndex = Number(e.target.parentNode.id.split('_')[1]);
        // ...
        renderTask(taskIndex);
      });
    });
  });

  deleteButtons.forEach((deleteButton) => {
    deleteButton.addEventListener('mousedown', (e) => {
      const taskIndex = Number(e.target.parentNode.id.split('_')[1]);
      taskManager.deleteTask(taskIndex);
      e.target.parentNode.remove();
    });
  });

  statuses.forEach((status) => {
    status.addEventListener('change', (e) => {
      const taskIndex = Number(e.target.parentNode.id.split('_')[1]);
      taskManager.modifyTask({ completed: e.target.checked }, taskIndex);
    });
  });
};
```
Instead of adding event listeners to each individual element, use event delegation to handle events at a higher level. This reduces the amount of code and simplifies the logic.
javascript
```
const addEventDelegationListeners = () => {
  // add event listener to the todo collection to handle events on todo items
  const todoCollection = document.getElementById('todo-collection');
  todoCollection.addEventListener('focus', (e) => {
    const todoEntry = e.target.parentNode;
    todoEntry.classList.add('highlight-input');
    todoEntry.querySelector('.delete').classList.remove('hide');
    todoEntry.querySelector('.reorder').classList.remove('hide');
  }, true);

  todoCollection.addEventListener('blur', (e) => {
    const todoEntry = e.target.parentNode;
    todoEntry.classList.remove
```